src/GraphBLAS-sharp.Backend/Common/ClArray.fs

```diff
@@ -437,6 +437,91 @@ module ClArray =
 
                 Some result
 
+    let assignWithKeysOption (op: Expr<'a -> 'b option>) (clContext: ClContext) workGroupSize =
+
+        let assign =
+            <@ fun (ndRange: Range1D) length (keys: ClArray<int>) (values: ClArray<'a>) (positions: ClArray<int>) (resultKeys: ClArray<int>) (resultValues: ClArray<'b>) resultLength ->
+
+                let gid = ndRange.GlobalID0
+
+                if gid < length then
+                    let position = positions.[gid]
+                    let value = values.[gid]
+                    let key = keys.[gid]
+
+                    // seems like scatter (option scatter) ???
+                    if 0 <= position && position < resultLength then
+                        match (%op) value with
+                        | Some value ->
+                            resultValues.[position] <- value
+                            resultKeys.[position] <- key
+
+                        | None -> () @>
+
+        let kernel = clContext.Compile assign
+
+        fun (processor: MailboxProcessor<_>) (keys: ClArray<int>) (values: ClArray<'a>) (positions: ClArray<int>) (resultKeys: ClArray<int>) (resultValues: ClArray<'b>) ->
+
+            if values.Length <> positions.Length then
+                failwith "lengths must be the same"
+
+            let ndRange =
+                Range1D.CreateValid(values.Length, workGroupSize)
+
+            let kernel = kernel.GetKernel()
+
+            processor.Post(
+                Msg.MsgSetArguments
+                    (fun () ->
+                        kernel.KernelFunc
+                            ndRange
+                            values.Length
+                            keys
+                            values
+                            positions
+                            resultKeys
+                            resultValues
+                            resultValues.Length)
+            )
+
+            processor.Post(Msg.CreateRunMsg<_, _>(kernel))
+
+    let chooseWithKeys<'a, 'b> (predicate: Expr<'a -> 'b option>) (clContext: ClContext) workGroupSize =
```

Comment:
I could not find any use of this function. What is it required for? If it is supposed to be applied to a `vector`, then wouldn't it be better to implement it in the `vector module`?

Reply:
It was previously used in SpMSpV, but later it's functionality was expressed through other functions so now this can be deleted